core/arch/arm/mm/core_mmu_v7.c

```diff
@@ -262,20 +269,66 @@ static paddr_t core_mmu_get_ul1_ttb_pa(struct mmu_partition *prtn)
 	return pa;
 }
 
-static void *core_mmu_alloc_l2(struct mmu_partition *prtn, size_t size)
+static uint32_t *alloc_l2_table(struct mmu_partition *prtn)
 {
-	uint32_t to_alloc = ROUNDUP(size, NUM_L2_ENTRIES * SMALL_PAGE_SIZE) /
-		(NUM_L2_ENTRIES * SMALL_PAGE_SIZE);
-
-	DMSG("L2 table used: %d/%d", prtn->tables_used + to_alloc,
-	     MAX_XLAT_TABLES);
-	if (prtn->tables_used + to_alloc > MAX_XLAT_TABLES)
-		return NULL;
-
-	memset(prtn->l2_tables[prtn->tables_used], 0,
-		sizeof(l2_xlat_tbl_t) * to_alloc);
-	prtn->tables_used += to_alloc;
-	return prtn->l2_tables[prtn->tables_used - to_alloc];
+	uint32_t *new_table = NULL;
+
+	/* The CFG_DYN_CONFIG implementation below depends on this */
+	static_assert(sizeof(l2_xlat_tbl_t) * 4 == SMALL_PAGE_SIZE);
+
+	if (IS_ENABLED(CFG_DYN_CONFIG)) {
+		void *p = NULL;
+
+		if (prtn->last_l2_page)
+			goto dyn_out;
+		if (cpu_mmu_enabled()) {
+			tee_mm_entry_t *mm = NULL;
+			paddr_t pa = 0;
+
+			mm = phys_mem_core_alloc(SMALL_PAGE_SIZE);
+			if (!mm) {
+				EMSG("Phys mem exhausted");
+				return NULL;
+			}
+			pa = tee_mm_get_smem(mm);
+
+			p = phys_to_virt(pa, MEM_AREA_SEC_RAM_OVERALL,
+					 SMALL_PAGE_SIZE);
+			assert(p);
+		} else {
+			p = boot_mem_alloc(SMALL_PAGE_SIZE, SMALL_PAGE_SIZE);
+			if (prtn->last_l2_page)
+				assert(prtn->last_l2_page + SMALL_PAGE_SIZE ==
+				       p);
+			if (!prtn->tables_used)
+				boot_mem_add_reloc(&prtn->last_l2_page);
+		}
+		prtn->last_l2_page = p;
+dyn_out:
+		new_table = (void *)(prtn->last_l2_page +
+				     (prtn->tables_used % 4) *
+				     sizeof(l2_xlat_tbl_t));
+		prtn->tables_used++;
+		/*
+		 * The current page is exhausted now, the next time we need
+		 * to allocate a new one.
+		 */
+		if (!(prtn->tables_used % 4))
+			prtn->last_l2_page = NULL;
+		DMSG("L2 tables used %u", prtn->tables_used);
+	} else {
+		if (prtn->tables_used >= MAX_XLAT_TABLES) {
+			EMSG("%u L2 tables exhausted", MAX_XLAT_TABLES);
+			return NULL;
+		}
+
+		new_table = prtn->l2_tables[prtn->tables_used];
+		prtn->tables_used++;
+		DMSG("L2 table used: %d/%d", prtn->tables_used,
```

Comment:
Nitpicking: typo in previous implementation, let's fix it?
`"L2 table used: %"PRIu32"/%d"`